### ISSUE TITLE
Deduplication of arrays of objects broken for data OUT

### ISSUE THREAD

**Issue by parisni** [tiangolo/fastapi#3017]
### First check

* [X] I added a very descriptive title to this issue.
* [X] I used the GitHub search to find a similar issue and didn't find it.
* [X] I searched the FastAPI documentation, with the integrated search.
* [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [X] I already read and followed all the tutorial in the docs and didn't find an answer.
* [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [X] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from pydantic import BaseModel
from typing import Set
from fastapi import FastAPI

app = FastAPI()


class Request(BaseModel):
    request_id: str

    def __hash__(self):
        return hash(self.request_id)


class Requests(BaseModel):
    request_ids: Set[Request]


@app.post("/test", response_model=Requests)
def read_root(data: Requests):
    return data

```

### Description

Thanks to the [Set feature](https://fastapi.tiangolo.com/tutorial/body-nested-models/?h=duplicate#set-types), the deduplication step works ok for data IN. however it fails for data OUT:

```bash
curl -X 'POST'   'http://127.0.0.1:8000/test'   -H 'accept: application/json'   -H 'Content-Type: application/json'   -d '{

  "request_ids": [
    {
      "request_id": "c04ca7ee-f62c-4372-b667-73ace53c69c2"
    }
  ]
}'

  File "pydantic/main.py", line 504, in pydantic.main.BaseModel.dict

  File "pydantic/main.py", line 869, in _iter
  File "pydantic/main.py", line 810, in pydantic.main.BaseModel._get_value
TypeError: unhashable type: 'dict'
```

### Environment

* OS: Ubuntu 20.04.2 LTS
* FastAPI Version [e.g. 0.3.0]: 0.63.0
* Python Version: 3.8.5

**Comment by Dustyposa** [comment]
You should change model `Requests` to:
```python
class Requests(BaseModel):
    request_ids: List[Set[Request]]

```

**Comment by parisni** [comment]
> You should change model `Requests` to:
> 
> ```python
> class Requests(BaseModel):
>     request_ids: List[Set[Request]]
> ```

I don't have two level of nested request. This does not solve the issue

**Comment by coder3112** [comment]
What about dedulpication in the code otself like so:
``` python
from pydantic import BaseModel
from typing import Set
from fastapi import FastAPI

app = FastAPI()


class Request(BaseModel):
    request_id: str

    def __hash__(self):
        return hash(self.request_id)


class Requests(BaseModel):
    request_ids: List[Request]


@app.post("/test", response_model=Requests)
def read_root(data: Requests):
    return list(set(data))
```

**Comment by parisni** [comment]
> What about de-duplication in the code otself like so:

that could be but from the doc linked bellow:
> And whenever you output that data, even if the source had duplicates, it will be output as a set of unique items.

so this is likely a BUG.

**Comment by coder3112** [comment]
> > And whenever you output that data, even if the source had duplicates, it will be output as a set of unique items.
> 
> so this is likely a BUG.
How is this a bug? You need a set, don't you? Your Pydantic Model as well as issue title suggests that

**Comment by parisni** [comment]
> If you mean that duplicates are not deduplicated:

nope. I meant an error is raised. please take a look at the example I gave.

**Comment by coder3112** [comment]
> nope. I meant an error is raised. please take a look at the example I gave.

Oh, sorry, I was using a custom version of FastAPI on my machine. Have changed a few things about serialization there so did not give error.
But on a fresh installation, the error I receive is:
```TypeError: unhashable type: 'dict'```
I am assuming this is the "error" you are talking about

In python, only immutable data types are hashable, so this was giving an error while being serialized by ```Pydantic```.

So, a workaround you might be able to use is:
```python
from pydantic import BaseModel
from typing import List
from fastapi import FastAPI

app = FastAPI()


class Request(BaseModel):
    request_id: str

    def __hash__(self):
        return hash(self.request_id)


class Requests(BaseModel):
    request_ids: List[Request]


@app.post("/test")
def test(data: Requests):
    data.request_ids = list(set(data.request_ids))
    return data
```
